### ISSUE TITLE
Json works, jsonb doesnt, may be due to the type mapper

### ISSUE THREAD

**Issue by yvesnyc** [tminglei/slick-pg#141]
Thanks for slick-pg.

It works for json tables. However, I've been unable to get the jsonb support to work. It seems the type mapper fails when using jsonb. 

For json I see the type map: com.github.tminglei.slickpg.utils.PgCommonJdbcTypes$Generic/json. 
For jsonb I see the type map: com.github.tminglei.slickpg.utils.PgCommonJdbcTypes$Generic/null. The Postgres driver then fails on the "null" type. Below is more data.

Here is my Driver which works when using json. Followed by the build.sbt configuration. I'm using slick-pg 0.84.

```
------------slickExtendedDriver----

import com.github.tminglei.slickpg._
import slick.driver.PostgresDriver

trait slickExtendedPostgresDriver extends PostgresDriver with PgJson4sSupport {

override lazy val Implicit = new ImplicitsPlus {}
override val simple = new SimpleQLPlus {}

// JsonB
// override val pgjson = "jsonb" //to keep back compatibility, pgjson's value was "json" by default

trait ImplicitsPlus extends Implicits with JsonImplicits
trait SimpleQLPlus extends SimpleQL with ImplicitsPlus
/// for json support
type DOCType = text.Document
override val jsonMethods = org.json4s.native.JsonMethods
}
object myDriverPostgresDriver extends slickExtendedPostgresDriver
-----------------------------------------
```

```
-------------build.sbt-----------------
 "org.json4s" %% "json4s-native" % "3.2.10",
  "org.json4s" %% "json4s-jackson" % "3.2.10",
  "org.json4s" %% "json4s-ast" % "3.2.10",
"com.typesafe.slick" %% "slick" % "2.1.0",
"org.postgresql" % "postgresql" % "9.4-1201-jdbc4",
"com.github.tminglei" %% "slick-pg" % "0.8.4"
-----------------------------------------
```

In debug mode I see messages from the slick compiler
```
11:37:32.842 TKD [main] DEBUG scala.slick.compiler.QueryCompiler - After phase assignUniqueSymbols:
| TableExpansion
| table s2: Table repository : Coll[scala.collection.immutable.Vector][NominalType(@(t3))(UnassignedStructuralType(@(t3)))]
| columns: TypeMapping
| 0: ProductNode
| 1: Path s2.url : String/VARCHAR(254)
| 2: Path s2.id : Int/INTEGER
| 3: Path s2.branch : Option[String/VARCHAR(254)]
| 4: Path s2.scmversion : com.github.tminglei.slickpg.utils.PgCommonJdbcTypes$Generic/json
```

When I use jsonb by uncommenting the 'override val pgjson = "jsonb"', I see the following debug messages. Notice the null at the end. 
```
11:28:07.993 TKD [main] DEBUG scala.slick.compiler.QueryCompiler - Source:
| TableExpansion
| table s2: Table repository : Coll[scala.collection.immutable.Vector][NominalType(@(PostgresDriver._.repository))(UnassignedStructuralType(@(PostgresDriver._.repository)))]
| columns: TypeMapping
| 0: ProductNode
| 1: Path s2.url : String/VARCHAR(254)
| 2: Path s2.id : Int/INTEGER
| 3: Path s2.branch : Option[String/VARCHAR(254)]
| 4: Path s2.scmversion : com.github.tminglei.slickpg.utils.PgCommonJdbcTypes$Generic/null
```

I even tried using json4s 3.2.11, but no change. 

Any Ideas?

Here is the Exception:

```
Exception in thread "main" org.postgresql.util.PSQLException: Unknown type null.
at org.postgresql.jdbc2.AbstractJdbc2Statement.setPGobject(AbstractJdbc2Statement.java:1703)
at org.postgresql.jdbc2.AbstractJdbc2Statement.setObject(AbstractJdbc2Statement.java:1926)
at org.postgresql.jdbc3g.AbstractJdbc3gStatement.setObject(AbstractJdbc3gStatement.java:36)
at org.postgresql.jdbc4.AbstractJdbc4Statement.setObject(AbstractJdbc4Statement.java:48)
at scala.slick.jdbc.JdbcBackend$SessionDef$LoggingPreparedStatement.setObject(JdbcBackend.scala:372)
at com.github.tminglei.slickpg.utils.PgCommonJdbcTypes$GenericJdbcType.setValue(PgCommonJdbcTypes.scala:23)
at scala.slick.jdbc.BaseResultConverter.set(JdbcResultConverter.scala:19)
at scala.slick.jdbc.BaseResultConverter.set(JdbcResultConverter.scala:11)
at scala.slick.relational.ProductResultConverter.set(ResultConverter.scala:71)
at scala.slick.relational.ProductResultConverter.set(ResultConverter.scala:46)
at scala.slick.relational.TypeMappingResultConverter.set(ResultConverter.scala:127)
at scala.slick.driver.JdbcInsertInvokerComponent$BaseInsertInvoker$$anonfun$internalInsert$1.apply(JdbcInsertInvokerComponent.scala:182)
at scala.slick.driver.JdbcInsertInvokerComponent$BaseInsertInvoker$$anonfun$internalInsert$1.apply(JdbcInsertInvokerComponent.scala:180)
at scala.slick.jdbc.JdbcBackend$SessionDef$class.withPreparedInsertStatement(JdbcBackend.scala:198)
at scala.slick.jdbc.JdbcBackend$BaseSession.withPreparedInsertStatement(JdbcBackend.scala:389)
at scala.slick.driver.JdbcInsertInvokerComponent$ReturningInsertInvoker.preparedInsert(JdbcInsertInvokerComponent.scala:301)
at scala.slick.driver.JdbcInsertInvokerComponent$BaseInsertInvoker.internalInsert(JdbcInsertInvokerComponent.scala:180)
at scala.slick.driver.JdbcInsertInvokerComponent$BaseInsertInvoker.insert(JdbcInsertInvokerComponent.scala:175)
at scala.slick.driver.JdbcInsertInvokerComponent$InsertInvokerDef$class.$plus$eq(JdbcInsertInvokerComponent.scala:70)
at scala.slick.driver.JdbcInsertInvokerComponent$BaseInsertInvoker.$plus$eq(JdbcInsertInvokerComponent.scala:145)
```

**Comment by tminglei** [comment]
Hi @yvesnyc, are you using PostgreSQL 9.4? `jsonb` wasn't supported before PostgreSQL 9.4.

**Comment by yvesnyc** [comment]
Yes, I am using PostgreSQL 9.4.1 (Amazon RDS). I created tables with ``jsonb``` type columns using a database client app. Not sure what else to look at.

**Comment by yvesnyc** [comment]
The driver does recognize ```jsonb```. I just tried it again without overriding ```pgjson=jsonb``` and here is the error message:

```
Exception in thread "main" org.postgresql.util.PSQLException: ERROR: column "scmversion" is of type jsonb but expression is of type json
  Hint: You will need to rewrite or cast the expression.
  Position: 139
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2270)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1998)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:255)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:570)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:420)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeUpdate(AbstractJdbc2Statement.java:366)
	at scala.slick.jdbc.JdbcBackend$SessionDef$LoggingPreparedStatement$$anonfun$executeUpdate$5.apply$mcI$sp(JdbcBackend.scala:331)
	at scala.slick.jdbc.JdbcBackend$SessionDef$LoggingPreparedStatement$$anonfun$executeUpdate$5.apply(JdbcBackend.scala:331)
	at scala.slick.jdbc.JdbcBackend$SessionDef$LoggingPreparedStatement$$anonfun$executeUpdate$5.apply(JdbcBackend.scala:331)
	at scala.slick.jdbc.JdbcBackend$SessionDef$LoggingStatement.logged(JdbcBackend.scala:264)
	at scala.slick.jdbc.JdbcBackend$SessionDef$LoggingPreparedStatement.executeUpdate(JdbcBackend.scala:331)
	at scala.slick.driver.JdbcInsertInvokerComponent$BaseInsertInvoker$$anonfun$internalInsert$1.apply(JdbcInsertInvokerComponent.scala:183)
	at scala.slick.driver.JdbcInsertInvokerComponent$BaseInsertInvoker$$anonfun$internalInsert$1.apply(JdbcInsertInvokerComponent.scala:180)
	at scala.slick.jdbc.JdbcBackend$SessionDef$class.withPreparedInsertStatement(JdbcBackend.scala:198)
	at scala.slick.jdbc.JdbcBackend$BaseSession.withPreparedInsertStatement(JdbcBackend.scala:389)
	at scala.slick.driver.JdbcInsertInvokerComponent$ReturningInsertInvoker.preparedInsert(JdbcInsertInvokerComponent.scala:301)
	at scala.slick.driver.JdbcInsertInvokerComponent$BaseInsertInvoker.internalInsert(JdbcInsertInvokerComponent.scala:180)
	at scala.slick.driver.JdbcInsertInvokerComponent$BaseInsertInvoker.insert(JdbcInsertInvokerComponent.scala:175)
	at scala.slick.driver.JdbcInsertInvokerComponent$InsertInvokerDef$class.$plus$eq(JdbcInsertInvokerComponent.scala:70)
	at scala.slick.driver.JdbcInsertInvokerComponent$BaseInsertInvoker.$plus$eq(JdbcInsertInvokerComponent.scala:145)
```

**Comment by tminglei** [comment]
Well, you need override `pgjson=jsonb`, to keep the type consistent with db column.

Let's continue to analyze the cause of **first** exception.

Here's the codes in `AbstractJdbc2Statement.java` of pg driver where throw the exception:
```java
    // Helper method for setting parameters to PGobject subclasses.
    private void setPGobject(int parameterIndex, PGobject x) throws SQLException {
        String typename = x.getType();
        int oid = connection.getTypeInfo().getPGType(typename);
        if (oid == Oid.UNSPECIFIED)
            throw new PSQLException(GT.tr("Unknown type {0}.", typename), PSQLState.INVALID_PARAMETER_TYPE);
...
```

And here's the codes in `TypeInfoCache.java` of pg driver that fetch the pg type:
```java
    public synchronized int getPGType(String pgTypeName) throws SQLException
    {
        Integer oid = (Integer)_pgNameToOid.get(pgTypeName);
        if (oid != null)
            return oid.intValue();

        if (_getOidStatement == null) {
            String sql;
            if (_conn.haveMinimumServerVersion("8.0")) {
                // see comments in @getSQLType()
                sql = "SELECT pg_type.oid " +
                      "  FROM pg_catalog.pg_type " +
                      "  LEFT " +
                      "  JOIN (select ns.oid as nspoid, ns.nspname, r.r " +
                      "          from pg_namespace as ns " +
                      "          join ( select s.r, (current_schemas(false))[s.r] as nspname " +
                      //                  -- go with older way of unnesting array to be compatible with 8.0
                      "                   from generate_series(1, array_upper(current_schemas(false), 1)) as s(r) ) as r " +
                      "         using ( nspname ) " +
                      "       ) as sp " +
                      "    ON sp.nspoid = typnamespace " +
                      " WHERE typname = ? " +
                      " ORDER BY sp.r, pg_type.oid DESC LIMIT 1;";
            } else if (_conn.haveMinimumServerVersion("7.3")) {
                sql = "SELECT oid FROM pg_catalog.pg_type WHERE typname = ? ORDER BY oid DESC LIMIT 1";
            } else {
                sql = "SELECT oid FROM pg_type WHERE typname = ? ORDER BY oid DESC LIMIT 1";
            }

            _getOidStatement = _conn.prepareStatement(sql);
        }

        _getOidStatement.setString(1, pgTypeName);

        // Go through BaseStatement to avoid transaction start.
        if (!((BaseStatement)_getOidStatement).executeWithFlags(QueryExecutor.QUERY_SUPPRESS_BEGIN))
            throw new PSQLException(GT.tr("No results were returned by the query."), PSQLState.NO_DATA);

        oid = new Integer(Oid.UNSPECIFIED);
        ResultSet rs = _getOidStatement.getResultSet();
        if (rs.next()) {
            oid = new Integer((int)rs.getLong(1));
            _oidToPgName.put(oid, pgTypeName);
        }
        _pgNameToOid.put(pgTypeName, oid);
        rs.close();

        return oid.intValue();
    }
```

I suspect no results returned for the sql, pls help confirm:
```
SELECT pg_type.oid , pg_type.*
FROM pg_catalog.pg_type 
   LEFT 
   JOIN (select ns.oid as nspoid, ns.nspname, r.r 
         from pg_namespace as ns
           join ( select s.r, (current_schemas(false))[s.r] as nspname 
           -- go with older way of unnesting array to be compatible with 8.0
                from generate_series(1, array_upper(current_schemas(false), 1)) as s(r) ) as r 
          using ( nspname ) 
         ) as sp
    ON sp.nspoid = typnamespace 
WHERE typname = 'jsonb'   -- this is the parameter
ORDER BY sp.r, pg_type.oid DESC;
```

**Comment by yvesnyc** [comment]
I ran the query on the 9.4.1 PostgreSQL DB. Here is the result with column names in the first row:

oid,typname,typnamespace,typowner,typlen,typbyval,typtype,typcategory,typispreferred,typisdefined,typdelim,typrelid,typelem,typarray,typinput,typoutput,typreceive,typsend,typmodin,typmodout,typanalyze,typalign,typstorage,typnotnull,typbasetype,typtypmod,typndims,typcollation,typdefaultbin,typdefault,typacl
3802,jsonb,11,10,-1,f,b,U,f,t,",",0,0,3807,jsonb_in,jsonb_out,jsonb_recv,jsonb_send,-,-,-,i,x,f,0,-1,0,0,,,

**Comment by tminglei** [comment]
Hi @yvesnyc, it's so strange to me. As you know, all of `slick-pg` json support tests are target to `jsonb` now, and they look ok.

Can you debug into the postgres driver to check? Is it possible that it connected to some other db?

--------------------------------------------------------------------------
_p.s. as I know, @arjunjun4u reported same problem in the comments of #138, but it was resolved by adjusting the mapped pg json type._

**Comment by yvesnyc** [comment]
The driver is connected to a database with jsonb columns. 

I tried it many times.

**Comment by arjunjun4u** [comment]
Hi @yvesnyc,

Not sure if it helps, but I had to explicitly define the dbtype while defining the fields (even for json)
```
def json = column[JValue]("json", O.DBType("json"))
```
The problem for me boiled down to mapping the correct field types - json vs jsonb. Not defining dbtype at all (the slick-pg examples doesn't define them) resulted in the same error as well.

See if adding  ```O.DBType("jsonb")``` helps your case.

**Comment by yvesnyc** [comment]
@arjunjun4u,

Thanks for the information but it didn't work for me. The default json setting worked without a problem for me. My issue came when I tried jsonb.

**Comment by yvesnyc** [comment]
I found the solution.

Slick-pg worked for me with the ```json``` type without a problem.
Slick-pg also supports ```jsonb```. To use ```jsonb``` you override the pgjson val in your driver.
``` override val pgjson = "jsonb"```

When I tried to use the ```jsonb``` type it wouldn't work (```Exception in thread "main" org.postgresql.util.PSQLException: Unknown type null.```). 
 
Under debug mode I compared messages for ```json``` and the failing ```jsonb``` and noticed the following.

```json``` - ``com.github.tminglei.slickpg.utils.PgCommonJdbcTypes$Generic/json```
```jsonb``` - ```com.github.tminglei.slickpg.utils.PgCommonJdbcTypes$Generic/null ```

Notice that the failure is due to a null as the JDBC type.

After lots of fix attempts I noticed I could override pgjson to any value, even ```"json"``` or, incredibly, ```""``` and I would get the same failure.

So I downloaded the Slick-pg source and changed the default pgjson to ```"jsonb"``` and did not override it. It worked!

So why would overriding a val cause an error or in this case a null. Googling this question led to this Stack page: http://stackoverflow.com/questions/16346929/scala-pre-initialize-val-while-extending-a-class?rq=1

Summary: under certain conditions/ordering Scala will set a val to null when overriding.

I solved my problem without altering the Slick-pg source. Simply define your driver like so:

```
//to keep back compatibility, pgjson's value was "json" by default, so override it here in the trait signature
trait slickExtendedPostgresDriver extends {override val pgjson="jsonb"} with PostgresDriver with PgJson4sSupport {

  override lazy val Implicit = new ImplicitsPlus {}
  override val simple = new SimpleQLPlus {}

  trait ImplicitsPlus extends Implicits  with JsonImplicits

  trait SimpleQLPlus extends SimpleQL with ImplicitsPlus

  type DOCType = text.Document

  override val jsonMethods = org.json4s.native.JsonMethods
}

```

I'll close the issue.

Thanks to all. And thanks for Slick-pg.

**Comment by tminglei** [comment]
_Glad to hear that. And I also learned something new. Thanks! :)_

**Comment by tminglei** [comment]
Hi @yvesnyc, I reproduced this problem when refactoring tests cases in `master` branch on local.
So, to be safer, I changed `val pgjson = "json"` to `def pgjson: String`.

I'll release it some later. Pls help check.

**Comment by yvesnyc** [comment]
Sounds good.

I’ll keep a look out for the release.

Thanks.

**Comment by tminglei** [comment]
Hi @yvesnyc, I released `slick-pg` v0.8.5, which included the changes.
Pls help check. _Thanks!_
